src/adapters/adbutler.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * @overview AdButler Prebid.js adapter.
+ * @author dkharton
+ */
+
+'use strict';
+
+var utils = require('../utils.js');
+var adloader = require('../adloader.js');
+var bidmanager = require('../bidmanager.js');
+var bidfactory = require('../bidfactory.js');
+
+var AdButlerAdapter = function AdButlerAdapter() {
+  
+  function _callBids(params) {
+    
+    var bids = params.bids || [],
+        callbackData = {},
+        zoneCount = {},
+        pageID = Math.floor(Math.random() * 10e6);
+
+    //Build and send bid requests
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i],
+        zoneID = utils.getBidIdParamater('zoneID', bid.params),
+        callbackID;  
+        
+      if(!(zoneID in zoneCount)){
+        zoneCount[zoneID] = 0;
+      }
+      
+      //build callbackID to get placementCode later  
+      callbackID = zoneID + '_' +zoneCount[zoneID];
+
+      callbackData[callbackID] = {};
+
+      callbackData[callbackID].placementCode = bid.placementCode;
+      callbackData[callbackID].sizes = bid.sizes;
+      callbackData[callbackID].maxCPM = utils.getBidIdParamater('maxCPM', bid.params);
+      callbackData[callbackID].minCPM = utils.getBidIdParamater('minCPM', bid.params);
+      
+      var adRequest = buildRequest(bid,zoneCount[zoneID],pageID);
+      zoneCount[zoneID]++;
+
+      adloader.loadScript(adRequest);
+    }
+      
+    //Define callback function for bid responses
+    $$PREBID_GLOBAL$$.adbutlerCB = function(aBResponseObject){
+        
+      var bidResponse = {},
+        callbackID = aBResponseObject.zone_id+'_'+aBResponseObject.place,
+        width = parseInt(aBResponseObject.width),
+        height = parseInt(aBResponseObject.height),
+        isCorrectSize = false,
+        isCorrectCPM = true,
+        CPM,minCPM,maxCPM;
+        
+      //Ensure that response ad matches one of the placement sizes.  
+      utils._each(callbackData[callbackID].sizes,function(size){
```

Comment:
This occasionally causes `Uncaught TypeError: Cannot read property 'sizes' of undefined` when running unit tests. Please check that `callbackData[callbackId]` exists before access the `sizes` property. See #662, #663, and #664 for PRs where we caught similar TypeErrors